# Predicting the Working Status (functional or not) of Pumps at Waterpoints in Tanzania

## Introduction to the Research Question

<p>For some communities in Africa, and here more specifically in Tanzania, having a secured access to fresh water is not a given. The number of waterpoints available to these communities is limited, and making sure that the pumps at these waterpoints are functional is critical.</p>
<p>The goal of this study is to find an algorithm to identify, or predict, which pumps are likely to be broken. This information is invaluable in order to improve maintenance operations and ensure continuous access to the water.</p>
<p>The data set that I will be using for this study is provided by
<a href="http://taarifa.org/">Taarifa</a> and the <a href="http://maji.go.tz/">Tanzanian Ministry of Water</a>. In order to be able to predict if a pump at a waterpoint is functional or not, I will need to identify the explanatory variables in the data set that are the most correlated with our response variable which is the pump working status. Some of the predictors that can be envisaged to be useful to predict the working status of a pump are the contruction year, how the waterpoint is managed, the waterpoint location, and many others.</p>

## Method

### Sample

Taarifa is an open source platform for the crowd sourced reporting and triaging of infrastructure related issues.
The sample data used comes from the Taarifa waterpoint dashboard, which aggregates data from the Tanzania Ministry of Water. Therefore the data set only contains data for Tanzanian waterpoints.

The data provides information on what kind of pump is operating, when it was installed, how it is managed, and other vatiables that can be used to predict the status of pumps at waterpoints. The full data available will be used for this study (sample size: 59400). However some of the variables will not be used if they are not correlated to the response variable.

In [2]:
import pandas

trainingData = pandas.read_csv("Training Set Values.csv")
#Display the number of observations in the full data set
numberOfObservations = len(trainingData)
print("Number of observations in full data set: " + str(numberOfObservations))

Number of observations in full data set: 59400


Add number of variables.  
Add the type of varibales (categorical/quantitative).  
Describe response variable and the values it can take.


The sample contains the following x variables:
<table>
    <tr>
        <th>Variable</th>
        <th>Type</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>amount_tsh</td>
        <td></td>
        <td>Total static head (amount water available to waterpoint)</td>
    </tr>
    <tr>
        <td>date_recorded</td>
        <td></td>
        <td>The date the row was entered</td>
    </tr>
    <tr>
        <td>funder</td>
        <td></td>
        <td>Who funded the well</td>
    </tr>
    <tr>
        <td>gps_height</td>
        <td></td>
        <td>Altitude of the well</td>
    </tr>
    <tr>
        <td>installer</td>
        <td></td>
        <td>Organization that installed the well</td>
    </tr>
    <tr>
        <td>longitude</td>
        <td></td>
        <td>GPS coordinate</td>
    </tr>
    <tr>
        <td>latitude</td>
        <td></td>
        <td>GPS coordinate</td>
    </tr>
    <tr>
        <td>wpt_name</td>
        <td></td>
        <td>Name of the waterpoint if there is one</td>
    </tr>
    <tr>
        <td>num_private</td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>basin</td>
        <td></td>
        <td>Geographic water basin</td>
    </tr>
    <tr>
        <td>subvillage</td>
        <td></td>
        <td>Geographic location</td>
    </tr>
    <tr>
        <td>region</td>
        <td></td>
        <td>Geographic location</td>
    </tr>
    <tr>
        <td>region_code</td>
        <td></td>
        <td>Geographic location (coded)</td>
    </tr>
    <tr>
        <td>district_code</td>
        <td></td>
        <td>Geographic location (coded)</td>
    </tr>
    <tr>
        <td>lga</td>
        <td></td>
        <td>Geographic location</td>
    </tr>
    <tr>
        <td>ward</td>
        <td></td>
        <td>Geographic location</td>
    </tr>
    <tr>
        <td>population</td>
        <td></td>
        <td>Population around the well</td>
    </tr>
    <tr>
        <td>public_meeting</td>
        <td></td>
        <td>True/False</td>
    </tr>
    <tr>
        <td>recorded_by</td>
        <td></td>
        <td>Group entering this row of data</td>
    </tr>
    <tr>
        <td>scheme_management</td>
        <td></td>
        <td>Who operates the waterpoint</td>
    </tr>
    <tr>
        <td>scheme_name</td>
        <td></td>
        <td>Who operates the waterpoint</td>
    </tr>
    <tr>
        <td>permit</td>
        <td></td>
        <td>If the waterpoint is permitted</td>
    </tr>
    <tr>
        <td>construction_year</td>
        <td></td>
        <td>Year the waterpoint was constructed</td>
    </tr>
    <tr>
        <td>extraction_type</td>
        <td></td>
        <td>The kind of extraction the waterpoint uses</td>
    </tr>
    <tr>
        <td>extraction_type_group</td>
        <td></td>
        <td>The kind of extraction the waterpoint uses</td>
    </tr>
    <tr>
        <td>extraction_type_class</td>
        <td></td>
        <td>The kind of extraction the waterpoint uses</td>
    </tr>
    <tr>
        <td>management</td>
        <td></td>
        <td>How the waterpoint is managed</td>
    </tr>
    <tr>
        <td>management_group</td>
        <td></td>
        <td>How the waterpoint is managed</td>
    </tr>
    <tr>
        <td>payment</td>
        <td></td>
        <td>What the water costs</td>
    </tr>
    <tr>
        <td>payment_type</td>
        <td></td>
        <td>What the water costs</td>
    </tr>
    <tr>
        <td>water_quality</td>
        <td></td>
        <td>The quality of the water</td>
    </tr>
    <tr>
        <td>quality_group</td>
        <td></td>
        <td>The quality of the water</td>
    </tr>
    <tr>
        <td>quantity</td>
        <td></td>
        <td>The quantity of water</td>
    </tr>
    <tr>
        <td>quantity_group</td>
        <td></td>
        <td>The quantity of water</td>
    </tr>
    <tr>
        <td>source</td>
        <td></td>
        <td>The source of the water</td>
    </tr>
    <tr>
        <td>source_type</td>
        <td></td>
        <td>The source of the water</td>
    </tr>
    <tr>
        <td>source_class</td>
        <td></td>
        <td>The source of the water</td>
    </tr>
    <tr>
        <td>waterpoint_type</td>
        <td></td>
        <td>The kind of waterpoint</td>
    </tr>
    <tr>
        <td>waterpoint_type_group</td>
        <td></td>
        <td>The kind of waterpoint</td>
    </tr>
 </table> 
 


### Measures
<p>After looking at the data dictionary some of the variables seem like good candidates (type of pump, water quality...) to predict if a pump is working, need repair, or is not even working.</p>
<p>The variable chosen are:</p>
* **funder** - Who funded the well
* **installer** - Organization that installed the well
* **basin** - Geographic water basin
* **region** - Geographic location
* **population** - Population around the well
* **public_meeting** - Indicates if the pump is in a public place :True/False
* **scheme_management** - Who operates the waterpoint
* **construction_year** - Year the waterpoint was constructed
* **extraction_type** - The kind of extraction the waterpoint uses
* **extraction_type_group** - The kind of extraction the waterpoint uses
* **extraction_type_class** - The kind of extraction the waterpoint uses
* **management** - How the waterpoint is managed
* **management_group** - How the waterpoint is managed
* **water_quality** - The quality of the water
* **source** - The source of the water
* **waterpoint_type** - The kind of waterpoint
* **waterpoint_type_group** - The kind of waterpoint


The data is provided in two data sets: "Training Set Values.csv" and "Training Set Labels.csv" that contains the response variable status_group.

In [3]:
responseData = pandas.read_csv("Training Set Labels.csv")
print("Number of observations in Training Set Labels.csv: " +
      str(len(responseData)))

Number of observations in Training Set Labels.csv: 59400


The response variable status_group is a categorical variable that can take the values: 
* functional - the waterpoint is operational and there are no repairs needed
* functional needs repair - the waterpoint is operational, but needs repairs
* non functional - the waterpoint is not operational

In order to perform some of the analysis, the two data set needs to be joined. This can be done using the id variable available in both data sets.

In [4]:
myData = trainingData.merge(responseData, on='id', how='outer')
print(len(myData))

59400


The distribution of status_group is (%):

In [5]:
print(myData['status_group'].value_counts(sort=False, normalize=True)*100)

functional                 54.308081
non functional             38.424242
functional needs repair     7.267677
Name: status_group, dtype: float64


Only 54.3% of the pumps do not need any kind a repair.

### Analysis
<p>The first step of the analysis will be to check if the explanatory variable chosen are associated with the response variable.</p>
<p>The response variable is categorical, therefore we will use:</p>
* Chi Square test of independance if the explanatory variable is categorical
* Chi Square test of independance if the explanatory variable is quantitavice, but the explanatory variable will need to be categorized first

<p>The null hypothesis for a chi-square test of independance is that there is no relationship between tge two categorical variables (they are independant). The alternate hypothesis is that there is a relationaship between the two categorical variables (they are not independant).</p>

In this data set there are many explanatory variables that are correlated with the functioning status of a pump, trying to build a linear model will not work. This is why it is necassary to employ machine learning technics.  
As I have already identified the variables that I want to include in the model, I have chosen a supervised model, the decision tree. The response variable is categorical, so the model is called a classification tree.

## Results

### Descriptives and Bi-variate

##### Public Meeting Place 

<p>I will start by testing the public meeting place responce variable.</p>

In [6]:
print(myData['public_meeting'].value_counts(sort=False, normalize=True)*100)

False     8.510101
True     85.877104
Name: public_meeting, dtype: float64


85% of the pumps are in a public place.

In [7]:
ct_public_meeting = pandas.crosstab(myData['public_meeting'],
                                    myData['status_group'])
print(ct_public_meeting.apply(lambda r: r*100/r.sum(), axis=1))
print(ct_public_meeting)

status_group    functional  functional needs repair  non functional
public_meeting                                                     
False            42.987141                 8.743818       48.269041
True             55.689949                 7.290584       37.019466
status_group    functional  functional needs repair  non functional
public_meeting                                                     
False                 2173                      442            2440
True                 28408                     3719           18884


It shows that when the pump is at a public meeting place, 55.6% of the pumps are properly functioning, when only 42.9% are fully functioning when not a public meeting place.
Next I run a Chi Square test of independance to validate if the variables are correlated (the chsi-square value, p value and expected counts are displayed).

In [8]:
import scipy.stats

cs_public_meeting = scipy.stats.chi2_contingency(ct_public_meeting)
print (cs_public_meeting)

(302.18246204959723, 2.4094177307380953e-66, 2, array([[  2757.23174473,    375.16239789,   1922.60585738],
       [ 27823.76825527,   3785.83760211,  19401.39414262]]))


The chi value is large (302) and the p value (2.4e-66) is very small, which confirms that we can reject the null hypothesys and the working status of a pump is correlated with the pump being in a public place or not. No post hoc tests are needed as the explanantory variable has only two categories. 

##### Source

In [9]:
print(myData['source'].value_counts(sort=False, normalize=True)*100)

rainwater harvesting     3.863636
dam                      1.104377
hand dtw                 1.471380
spring                  28.654882
lake                     1.287879
unknown                  0.111111
river                   16.181818
machine dbh             18.644781
other                    0.356902
shallow well            28.323232
Name: source, dtype: float64


To be able to perfrm some of the analyses of this study, I need to convert the strings used in this categorical variable in integer.

In [10]:
source_map = {'rainwater harvesting':0,'spring':1,'lake':2,
              'shallow well':3,'unknown':4,'hand dtw':5,
              'other':6,'river':7,'dam':8,'machine dbh':9
             }
myData['source_c'] = myData['source'].map(source_map)
print(myData['source_c'].value_counts(sort=False, normalize=True)*100)

0     3.863636
1    28.654882
2     1.287879
3    28.323232
4     0.111111
5     1.471380
6     0.356902
7    16.181818
8     1.104377
9    18.644781
Name: source_c, dtype: float64


In [11]:
ct_source = pandas.crosstab(myData['source'], myData['status_group'])
print(ct_source.apply(lambda r: r*100/r.sum(), axis=1))
#print(ct_source)

status_group          functional  functional needs repair  non functional
source                                                                   
dam                    38.567073                 3.658537       57.774390
hand dtw               56.864989                 1.945080       41.189931
lake                   21.176471                 1.568627       77.254902
machine dbh            48.957111                 4.433409       46.609481
other                  59.433962                 0.471698       40.094340
rainwater harvesting   60.392157                13.681917       25.925926
river                  56.856013                12.702871       30.441115
shallow well           49.476938                 5.688302       44.834760
spring                 62.229011                 7.496622       30.274367
unknown                48.484848                 6.060606       45.454545


In [12]:
cs_source = scipy.stats.chi2_contingency(ct_source)
print (cs_source)

(2623.9982801502629, 0.0, 18, array([[  3.56261010e+02,   4.76759596e+01,   2.52063030e+02],
       [  4.74652626e+02,   6.35194949e+01,   3.35827879e+02],
       [  4.15456818e+02,   5.55977273e+01,   2.93945455e+02],
       [  6.01461995e+03,   8.04895202e+02,   4.25548485e+03],
       [  1.15133131e+02,   1.54074747e+01,   8.14593939e+01],
       [  1.24637045e+03,   1.66793182e+02,   8.81836364e+02],
       [  5.22009273e+03,   6.98569091e+02,   3.69333818e+03],
       [  9.13679152e+03,   1.22271394e+03,   6.46449455e+03],
       [  9.24377843e+03,   1.23703126e+03,   6.54019030e+03],
       [  3.58433333e+01,   4.79666667e+00,   2.53600000e+01]]))


The chi value is large (2623) and the p value (0.0) is very small, which confirms that we can reject the null hypothesys and the working status of a pump is correlated with the source of the water.  
The explanatory variable has more than two values, so a post hoc test is required to identify which categories...
Bonferroni adjustement:

##### Permit

In [13]:
print(myData['permit'].value_counts(sort=False, normalize=True)*100)
ct_permit = pandas.crosstab(myData['permit'], myData['status_group'])
print(ct_permit.apply(lambda r: r*100/r.sum(), axis=1))
cs_permit = scipy.stats.chi2_contingency(ct_permit)
print (cs_permit)

False    29.447811
True     65.407407
Name: permit, dtype: float64
status_group  functional  functional needs repair  non functional
permit                                                           
False          51.709353                 7.546307       40.744340
True           55.443735                 6.941728       37.614537
(67.790140875342928, 1.9035221956985326e-15, 2, array([[  9495.42652279,   1247.07802073,   6749.49545648],
       [ 21090.57347721,   2769.92197927,  14991.50454352]]))


##### Construction year - Age
The date_recorded variable gives us the date at which the row (data) was recorded. The construction_year variable gives the year at which the waterpoint was constructed, but not the exact date. Therefore the age of the water point is approximated to: year(date_recorded) - construction_year.

In [14]:
import math
#Create scondary variable age.
def getAge (row):
    if math.isnan(row['construction_year']) or (row['construction_year'] == 0):
        return math.nan
    age = int(row['date_recorded'][:4]) - int(row['construction_year'])
    if (age >= 0):
        return age
    else:
        return math.nan

myData['waterpoint_age'] = myData.apply (lambda row: getAge(row),axis=1)
myData['waterpoint_age'].value_counts(sort=False, normalize=True)*100

0     0.989899
1     3.877104
2     3.584175
3     4.612795
4     3.181818
5     3.333333
6     2.324916
7     2.363636
8     1.952862
9     1.370370
10    1.461279
11    2.276094
12    0.941077
13    3.146465
14    1.952862
15    1.671717
16    1.390572
17    1.070707
18    1.269360
19    1.289562
20    0.910774
21    0.973064
22    0.543771
23    1.523569
24    0.439394
25    0.981481
26    0.991582
27    1.282828
28    1.193603
29    0.949495
30    0.604377
31    1.193603
32    0.292929
33    1.885522
34    0.309764
35    1.089226
36    0.547138
37    1.010101
38    0.590909
39    1.011785
40    0.392256
41    0.821549
42    0.151515
43    0.599327
44    0.079125
45    0.111111
46    0.080808
47    0.023569
48    0.069024
49    0.042088
50    0.141414
51    0.052189
52    0.018519
53    0.153199
Name: waterpoint_age, dtype: float64

To perform a chi-square test of independance, the explanatory variable needs to be a categorical variable. I decided to fold the waterpoint_age variable in 4 groups: < 5 years, 5-10 years, 10 to 20 years, 20-30 years, 30+.


In [15]:
#Let's create a secondary variable that will contain age groupings
myData["waterpoint_age_c"] = pandas.cut(
    myData.waterpoint_age,[0, 5, 10, 20, 30, 99],
             labels=["Group -5","Group 6-10","Group 11-20",
                     "Group 12-30", "group 30+"])
print(myData['waterpoint_age_c'].value_counts(sort=False, normalize=True)*100)

Group -5       18.589226
Group 6-10      9.473064
Group 11-20    15.919192
Group 12-30     9.483165
group 30+      10.666667
dtype: float64


In [16]:
ct_construction_year = pandas.crosstab(
    myData['waterpoint_age_c'], myData['status_group'])
print(ct_construction_year.apply(lambda r: r*100/r.sum(), axis=1))
cs_construction_year = scipy.stats.chi2_contingency(ct_construction_year)
print (cs_construction_year)

status_group      functional  functional needs repair  non functional
waterpoint_age_c                                                     
Group -5           71.599348                 4.673066       23.727586
Group 6-10         61.275991                 8.157100       30.566910
Group 11-20        57.614213                 6.694162       35.691624
Group 12-30        44.274809                 7.385052       48.340138
group 30+          31.423611                 7.780934       60.795455
(3169.8051407479829, 0.0, 8, array([[ 6170.29070195,   729.58245393,  4142.12684412],
       [ 3144.37835355,   371.79500709,  2110.82663937],
       [ 5284.030871  ,   624.79004568,  3547.17908332],
       [ 3147.73116501,   372.19144747,  2113.07738752],
       [ 3540.56890849,   418.64104583,  2376.79004568]]))


The crosstab analysis clearly shows that the older the waterpoint is, the less likely it is to be functional, there is a negative relationship. The chi-square anaylis confirms the correlation, the p value is 0.0.

#### Waterpoint type


In [17]:
print(myData['waterpoint_type'].value_counts(sort=False, normalize=True)*100)

dam                             0.011785
hand pump                      29.441077
communal standpipe multiple    10.274411
communal standpipe             48.016835
other                          10.740741
improved spring                 1.319865
cattle trough                   0.195286
Name: waterpoint_type, dtype: float64


In [18]:
ct_waterpoint_type = pandas.crosstab(
    myData['waterpoint_type'], myData['status_group'])
print(ct_waterpoint_type.apply(lambda r: r*100/r.sum(), axis=1))
cs_waterpoint_type = scipy.stats.chi2_contingency(ct_waterpoint_type)
print (cs_waterpoint_type)

status_group                 functional  functional needs repair  \
waterpoint_type                                                    
cattle trough                 72.413793                 1.724138   
communal standpipe            62.148517                 7.923708   
communal standpipe multiple   36.621334                10.617729   
dam                           85.714286                 0.000000   
hand pump                     61.785224                 5.884035   
improved spring               71.811224                10.841837   
other                         13.166144                 4.592476   

status_group                 non functional  
waterpoint_type                              
cattle trough                     25.862069  
communal standpipe                29.927775  
communal standpipe multiple       52.760937  
dam                               14.285714  
hand pump                         32.330741  
improved spring                   17.346939  
other              

To be able to perform some of the analyses of this study, I need to convert the strings used in this categorical variable in integer.

In [19]:
waterpoint_type_map = {'improved spring':0,'communal standpipe multiple':1,
                       'communal standpipe':2, 'cattle trough':3,
                       'other':4,'hand pump':5,'dam':6
             }
myData['waterpoint_type_c'] = myData['waterpoint_type'].map(waterpoint_type_map)
print(myData['waterpoint_type_c'].value_counts(sort=False, normalize=True)*100)

0     1.319865
1    10.274411
2    48.016835
3     0.195286
4    10.740741
5    29.441077
6     0.011785
Name: waterpoint_type_c, dtype: float64


### Multivariable

The variables set as my predictors are: the source of water, the type of meeting place (public or not), if the waterpoint was authorized, the waterpoint age, and the type of waterpoint.  
The data set is spilt into a training and test data sets, with 60% of the observations being used to train the model.


In [20]:
from sklearn.cross_validation import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report
import sklearn.metrics

In [21]:
#drop NA from the data set as decision tree analyses cannot handle any ANs
myData_clean = myData.dropna()

#Select the predictors that will be used
predictors = myData_clean[['source_c','public_meeting',
                           'permit','waterpoint_age','waterpoint_type_c']]

#Select the target, i.e. our response variable
targets = myData_clean.status_group

#Create the train and test data sets
pred_train, pred_test, tar_train, tar_test = train_test_split(
    predictors, targets, test_size=.4)

print(pred_train.shape)
print(pred_test.shape)
print(tar_train.shape)
print( tar_test.shape)

#Build model on training data
classifier=DecisionTreeClassifier()
classifier=classifier.fit(pred_train,tar_train)

predictions=classifier.predict(pred_test)

print(sklearn.metrics.confusion_matrix(tar_test,predictions))
print(sklearn.metrics.accuracy_score(tar_test, predictions))

(12972, 5)
(8649, 5)
(12972,)
(8649,)
[[4774   52  490]
 [ 423  100   99]
 [1398   77 1236]]
0.706440050873


The analyses shows that once trained, the decision tree is able to predict 70% of the working status of the waterpoints of the test data set.  
This is an encouraging result, but that can probably be improved by adding new explanatory variable to the model.